File: discord-demibot/src/http/routes/events.js
const express = require('express');
const send = require('../../discord/send');

module.exports = ({ db, discord, logger }) => {
  const router = express.Router();

  function buildEmbed(data, info, opts = {}) {
    const {
      title,
      description,
      time,
      color,
      url,
      fields,
      thumbnailUrl,
      authorName,
      authorIconUrl
    } = data;
    return {
      title,
      description,
      url,
      timestamp: time ? new Date(time) : undefined,
      color,
      footer: info.character ? { text: info.character } : undefined,
      fields,
      thumbnail: thumbnailUrl ? { url: thumbnailUrl } : undefined,
      author: authorName ? { name: authorName, icon_url: authorIconUrl } : undefined,
      image: opts.image ? { url: 'attachment://image.png' } : undefined
    };
  }

  router.post('/', async (req, res) => {
    const info = req.apiKey;
    const { channelId, title, time, description, imageBase64, color, url, fields, thumbnailUrl, authorName, authorIconUrl } = req.body;
    try {
      const embed = buildEmbed({ title, description, time, color, url, fields, thumbnailUrl, authorName, authorIconUrl }, info, { image: !!imageBase64 });
      const files = imageBase64 ? [{ attachment: Buffer.from(imageBase64, 'base64'), name: 'image.png' }] : [];
      const message = await send.send(channelId, embed, files);
      await db.addEventChannel(channelId);
      discord.trackEventChannel(channelId);
      const id = await db.saveEvent({
        userId: info.userId,
        channelId,
        messageId: message.id,
        title,
        description,
        time,
        metadata: imageBase64 ? 'image' : null
      });
      res.json({ ok: true, id });
    } catch (err) {
      if (logger) logger.error('Event creation failed', err);
      res.status(500).json({ ok: false });
    }
  });

  router.patch('/:id', async (req, res) => {
    const info = req.apiKey;
    const id = parseInt(req.params.id, 10);
    try {
      const existing = await db.getEvent(id);
      if (!existing) return res.status(404).json({ error: 'Not found' });
      if (existing.user_id !== info.userId && !info.isAdmin) return res.status(403).json({ error: 'Forbidden' });
      const title = req.body.title ?? existing.title;
      const description = req.body.description ?? existing.description;
      const time = req.body.time ?? existing.time;
      const imageBase64 = req.body.imageBase64;
      const color = req.body.color;
      const url = req.body.url;
      const fields = req.body.fields;
      const thumbnailUrl = req.body.thumbnailUrl;
      const authorName = req.body.authorName;
      const authorIconUrl = req.body.authorIconUrl;
      const embed = buildEmbed({ title, description, time, color, url, fields, thumbnailUrl, authorName, authorIconUrl }, info, { image: !!imageBase64 });
      const files = imageBase64 ? [{ attachment: Buffer.from(imageBase64, 'base64'), name: 'image.png' }] : [];
      await send.edit(existing.channel_id, existing.message_id, embed, files);
      await db.updateEvent({ id, title, description, time, metadata: imageBase64 ? 'image' : existing.metadata });
      res.json({ ok: true });
    } catch (err) {
      if (logger) logger.error('Event update failed', err);
      res.status(500).json({ ok: false });
    }
  });

  router.delete('/:id', async (req, res) => {
    const info = req.apiKey;
    const id = parseInt(req.params.id, 10);
    try {
      const existing = await db.getEvent(id);
      if (!existing) return res.status(404).json({ error: 'Not found' });
      if (existing.user_id !== info.userId && !info.isAdmin) return res.status(403).json({ error: 'Forbidden' });
      const title = existing.title.endsWith(' (Canceled)') ? existing.title : `${existing.title} (Canceled)`;
      const embed = buildEmbed({ title, description: existing.description, time: existing.time, color: 0x808080 }, info);
      await send.edit(existing.channel_id, existing.message_id, embed, []);
      await db.updateEvent({ id, title, description: existing.description, time: existing.time, metadata: 'canceled' });
      res.json({ ok: true });
    } catch (err) {
      if (logger) logger.error('Event delete failed', err);
      res.status(500).json({ ok: false });
    }
  });

  return router;
};

File: python_demibot/routes/events.py
import base64
import io
from datetime import datetime
from typing import Dict, Any

import discord
from fastapi import APIRouter, Depends, HTTPException

from ..api import get_api_key_info, db, bot
from python_demibot.rate_limiter import enqueue

router = APIRouter(prefix="/api/events")


class AttendanceButton(discord.ui.Button):
    """Button used for tracking event attendance."""

    def __init__(self, status: str):
        label = status.title()
        styles = {
            "yes": discord.ButtonStyle.success,
            "maybe": discord.ButtonStyle.secondary,
            "no": discord.ButtonStyle.danger,
        }
        super().__init__(label=label, style=styles[status], custom_id=f"attendance:{status}")
        self.status = status

    async def callback(self, interaction: discord.Interaction):  # pragma: no cover - depends on Discord
        event = await db.get_event_by_message_id(interaction.message.id)
        if event:
            await db.set_event_attendance(event["id"], str(interaction.user.id), self.status)
        await interaction.response.defer()


class AttendanceView(discord.ui.View):
    def __init__(self, buttons: list[str]):
        super().__init__(timeout=None)
        for status in buttons:
            self.add_item(AttendanceButton(status))


def build_embed(data: Dict[str, Any], info: Dict[str, Any], include_image: bool = False) -> discord.Embed:
    embed = discord.Embed(
        title=data.get("title"),
        description=data.get("description"),
        url=data.get("url"),
        color=data.get("color"),
    )
    if data.get("time"):
        try:
            embed.timestamp = datetime.fromisoformat(data["time"])
        except Exception:
            try:
                embed.timestamp = datetime.fromtimestamp(float(data["time"]) / 1000)
            except Exception:
                pass
    if info.get("characterName"):
        embed.set_footer(text=info["characterName"])
    for field in data.get("fields") or []:
        embed.add_field(name=field.get("name"), value=field.get("value"), inline=field.get("inline", False))
    if data.get("thumbnailUrl"):
        embed.set_thumbnail(url=data["thumbnailUrl"])
    if data.get("authorName"):
        embed.set_author(name=data.get("authorName"), icon_url=data.get("authorIconUrl"))
    if data.get("imageUrl"):
        embed.set_image(url=data["imageUrl"])
    elif include_image:
        embed.set_image(url="attachment://image.png")
    return embed


@router.post("/")
async def create_event(payload: Dict[str, Any], info: Dict[str, Any] = Depends(get_api_key_info)):
    channel_id = payload.get("channelId")
    try:
        channel = await bot.get_client().fetch_channel(channel_id)
        if not channel or channel.guild.id != int(info["serverId"]):
            raise HTTPException(status_code=403, detail={"ok": False})
        embed = build_embed(payload, info, include_image=bool(payload.get("imageBase64")))
        files = []
        if payload.get("imageBase64"):
            data = base64.b64decode(payload["imageBase64"])
            files = [discord.File(io.BytesIO(data), filename="image.png")]
        buttons = payload.get("attendance") or ["yes", "maybe", "no"]
        view = AttendanceView(buttons)
        message = await enqueue(lambda: channel.send(embed=embed, files=files, view=view))
        settings = await db.get_server_settings(info["serverId"])
        events = set(settings.get("eventChannels", []))
        events.add(channel_id)
        await db.update_server_settings(info["serverId"], {"eventChannels": list(events)})
        bot.track_event_channel(channel_id)
        event_id = await db.save_event({
            "userId": info["userId"],
            "channelId": channel_id,
            "messageId": str(message.id),
            "title": payload.get("title"),
            "description": payload.get("description"),
            "time": payload.get("time"),
            "metadata": "image" if payload.get("imageBase64") else None,
        })
        return {"ok": True, "id": event_id}
    except HTTPException:
        raise
    except Exception as err:  # pragma: no cover - defensive
        raise HTTPException(status_code=500, detail={"ok": False}) from err


@router.patch("/{event_id}")
async def update_event(event_id: int, payload: Dict[str, Any], info: Dict[str, Any] = Depends(get_api_key_info)):
    try:
        existing = await db.get_event(event_id)
        if not existing:
            raise HTTPException(status_code=404, detail={"error": "Not found"})
        channel = await bot.get_client().fetch_channel(existing["channel_id"])
        if not channel or channel.guild.id != int(info["serverId"]):
            raise HTTPException(status_code=403, detail={"error": "Forbidden"})
        if str(existing["user_id"]) != info["userId"] and not info.get("isAdmin"):
            raise HTTPException(status_code=403, detail={"error": "Forbidden"})
        data = {
            "title": payload.get("title", existing["title"]),
            "description": payload.get("description", existing["description"]),
            "time": payload.get("time", existing["time"]),
            "color": payload.get("color"),
            "url": payload.get("url"),
            "fields": payload.get("fields"),
            "thumbnailUrl": payload.get("thumbnailUrl"),
            "authorName": payload.get("authorName"),
            "authorIconUrl": payload.get("authorIconUrl"),
            "imageUrl": payload.get("imageUrl"),
        }
        image_base64 = payload.get("imageBase64")
        embed = build_embed(data, info, include_image=bool(image_base64))
        files = []
        if image_base64:
            file_bytes = base64.b64decode(image_base64)
            files = [discord.File(io.BytesIO(file_bytes), filename="image.png")]
        message = await channel.fetch_message(existing["message_id"])
        await enqueue(lambda: message.edit(embed=embed, attachments=files))
        await db.update_event({
            "id": event_id,
            "title": data["title"],
            "description": data["description"],
            "time": data["time"],
            "metadata": "image" if image_base64 else existing.get("metadata"),
        })
        return {"ok": True}
    except HTTPException:
        raise
    except Exception as err:  # pragma: no cover - defensive
        raise HTTPException(status_code=500, detail={"ok": False}) from err


@router.delete("/{event_id}")
async def delete_event(event_id: int, info: Dict[str, Any] = Depends(get_api_key_info)):
    try:
        existing = await db.get_event(event_id)
        if not existing:
            raise HTTPException(status_code=404, detail={"error": "Not found"})
        channel = await bot.get_client().fetch_channel(existing["channel_id"])
        if not channel or channel.guild.id != int(info["serverId"]):
            raise HTTPException(status_code=403, detail={"error": "Forbidden"})
        if str(existing["user_id"]) != info["userId"] and not info.get("isAdmin"):
            raise HTTPException(status_code=403, detail={"error": "Forbidden"})
        title = existing["title"]
        if not title.endswith(" (Canceled)"):
            title = f"{title} (Canceled)"
        embed = build_embed({
            "title": title,
            "description": existing["description"],
            "time": existing["time"],
            "color": 0x808080,
        }, info)
        message = await channel.fetch_message(existing["message_id"])
        await enqueue(lambda: message.edit(embed=embed, attachments=[]))
        await db.update_event({
            "id": event_id,
            "title": title,
            "description": existing["description"],
            "time": existing["time"],
            "metadata": "canceled",
        })
        return {"ok": True}
    except HTTPException:
        raise
    except Exception as err:  # pragma: no cover - defensive
        raise HTTPException(status_code=500, detail={"ok": False}) from err
